lab1_glebosotov

glebosotov opened the pull request:
@azazkamaz @igooor-bb @Markovvn1 @yadariya @tasneem22

glebosotov committed:
Lab submission

markovvn1 commented on app_python/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+WORKDIR /app
+RUN pip install -r requirements.txt
+ENTRYPOINT ["python3"]
+CMD ["app.py"]
```
The assignment says to use a production ready framework to run the server (you are running application in a development server). You probably want to use something like [gunicorn](https://pypi.org/project/gunicorn/) or [waitress](https://flask.palletsprojects.com/en/2.2.x/tutorial/deploy/#run-with-a-production-server)

markovvn1 commented on app_python/PYTHON.md:
```diff
@@ -0,0 +1,31 @@
+
+* add development files, that are nmot supposed to be pushed to .gitignore
+* use Docker for development and production
+* use virtualenv for development
```
You also use Jinja for templates, which is good practice

markovvn1 commented on app_flutter/moscow_time/lib/main.dart:
```diff
@@ -0,0 +1,52 @@
+              children: [
+                Text('Time in Moscow, Russia',
+                    style: Theme.of(context).textTheme.headline3),
+                Text(DateFormat('MM/dd/yyyy hh:mm:ss').format(DateTime.now()),
```
It seems like it has some problems with timezones. Right now is 09/04/2022 20:48:14 (not 08:48:14).
![image](https://user-images.githubusercontent.com/32509100/188326757-65c1558a-db23-4043-a6c3-b3311d32de4a.png)

AzazKamaz commented on app_flutter/moscow_time/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+#Stage 1 - Install dependencies and build the app in a build environment
+FROM debian:latest AS build-env
```
Better to use pre build flutter image. Building own each time takes too much time in production environment

glebosotov replied:
https://hub.docker.com/repository/docker/glebosotov/app_flutter_web

AzazKamaz commented on app_flutter/docker-compose.yml:
```diff
@@ -0,0 +1,8 @@
+    ports:
+      - "8080:80"
+    volumes:
+      - .moscow_time:/moscow_time
```
Volume seems not to be necessary for production given there is just nginx which serves /usr/share/nginx/html

AzazKamaz commented on app_flutter/moscow_time/test/widget_test.dart:
```diff
@@ -0,0 +1,30 @@
+import 'package:moscow_time/main.dart';
+
+void main() {
+  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
```
Test not passing

AzazKamaz commented on app_python/app.py:
```diff
@@ -0,0 +1,21 @@
+    return render_template('template.html', datetime = str(datetime.now(tz=tz)))
+
+if __name__ == '__main__':
+    app.run(debug=True, host='0.0.0.0')
```
debug=True is not production ready probably

glebosotov replied:
Resolved by using, `gunicorn` for running the server

igooor-bb commented on app_python/app.py:
```diff
@@ -0,0 +1,21 @@
+
+offset = timedelta(hours=3)
+tz = timezone(offset, name='MSK')
+
```
I believe linter requires 2 empty lines according to pep8

glebosotov committed:
Fix all issued based on PR review

- Fix default deployment of python web app
- Remove auto-generated tests from Flutter
- Fix timezone in Flutter
- Upload image to DockerHub for fast deployment

glebosotov commented:
Updated with all mentioned problems fixed 🙌🏻
